hive: Add support for group of cells and decorators

joamaki opened the pull request:
This commit is preparing for being able to have a custom "nested" lifecycle (see WithCustomLifecycle) in the operator so that we can have a subset of cells that are only started when the operator is elected leader.

--- 
This refactors the hive package in order to support "decorator" cells that wrap some set of cells to provide the set with an augmented object.

As it is looking more probable that it makes sense to implement directly on top of dig, this commit drops use of fx.Option directly and instead implements Provide and Invoke that take the functions directly. Additionally type aliases are provided for fx.Hook and fx.Lifecycle. This way transitioning towards a lighter implementation directly on top of dig is easier.

The cell related functions have been moved into "pkg/hive/cell" for cleaner separation from hive functions. Changes to other packages other than pkg/hive are to convert them to the new import path and shorter function names (NewCell->Provide).

Config is now a separate "cell type" and takes the default configuration as parameter. This allows a cleaner way for providing defaults to a cell (see gops.Cell).

The configuration dumping in "cilium-agent objects" is for now removed as the output was not very pretty and it is yet unclear how this should be done cleanly with nested cells.

joamaki committed:
hive: Add support for group of cells and decorators

This refactors the hive package in order to support "decorator" cells
that wrap a set of cells and provides them with augmented set of objects.

As it is looking more probable that it makes sense to implement directly on
top of dig, this commit drops use of fx.Option directly and instead implements
Provide and Invoke that take the functions directly. Additionally type aliases
are provided for fx.Hook and fx.Lifecycle. This way transitioning towards a lighter
implementation directly on top of dig is easier.

The cell related functions have been moved into "pkg/hive/cell" for cleaner
separation from hive functions. Changes to other packages other than pkg/hive
are to convert them to the new import path and shorter function names (NewCell->Provide).

Config is now a separate "cell type" and takes the default configuration as parameter.
This allows a cleaner way for providing defaults to a cell (see gops.Cell).

The configuration dumping in "cilium-agent objects" is for now removed as the
output was not very pretty and it is yet unclear how this should be done cleanly
with nested cells.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -1566,7 +1567,12 @@ func registerDaemonHooks(lc fx.Lifecycle, shutdowner fx.Shutdowner, cfg DaemonCe
+	// Set the global LocalNodeStore. This is to retain the API of getters and setters
+	// defined in pkg/node/address.go until uses of them have been converted to use
+	// LocalNodeStore directly.
+	node.SetLocalNodeStore(localNode)
```
I don't understand this diff. It seems separate from the refactoring of the hive / cells, no?

joamaki replied:
It made sense to move it here from `root.go`. Why this ended up in this PR was because I had implicitly relied on ordering of the invokes and when I accidently moved the `registerDaemonHooks` to be before the `SetLocalNodeStore` invoke in `root.go` things broke (the LocalNodeStore wasn't started before daemon as daemon didn't depend on it). I'll break it into its own commit.

christarazi reviewed (commented):
To be honest, I'm not really sure how to review the hive / cells code because I'm not familiar with the fx library. I understand what fx / hive / cells are trying to do at a high-level, but I feel like I can't really provide a proper review of the implementation. What can we do about this?

joamaki committed:
daemon: Move SetLocalNodeStore to registerDaemonHooks for correct order

We need to make sure that LocalNodeStore gets started before registerDaemonHooks
as otherwise Get/Update blocks. Fix this by making registerDaemonHooks depend
on LocalNodeStore and move the setting of node.localNode to it.

Since the list of parameters for registerDaemonHooks is getting quite long,
add daemonParams struct.

Fixes: d62bd53fad ("node: Add LocalNodeStore and convert getters/setters to it")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink commented on daemon/cmd/daemon_main.go:
```diff
@@ -1556,7 +1556,7 @@ func (d *Daemon) initKVStore() {
 // If an object still owned by Daemon is required in a module, it should be provided indirectly, e.g. via
 // a callback.
-func registerDaemonHooks(lc fx.Lifecycle, shutdowner fx.Shutdowner, cfg DaemonCellConfig, clientset k8sClient.Clientset) error {
+func registerDaemonHooks(lc fx.Lifecycle, shutdowner fx.Shutdowner, cfg DaemonCellConfig, localNodeStore node.LocalNodeStore, clientset k8sClient.Clientset) error {
```
This param list is getting fairly big, would it make sense to give it its own cell.In param struct?

joamaki replied:
Yep it does. I've done that in another PR (https://github.com/cilium/cilium/pull/21512/files#diff-c0e8dc92b48dcab08b42d07a5223eb03c267c8484f975a66f423ae85d9cc6ea2R1547), but I suppose I could introduce that change here already.

joamaki replied:
Done.

dylandreimerink commented on pkg/hive/hive.go:
```diff
@@ -42,10 +42,16 @@ const (
-// Hive is a modular application built from cells.
+// Hive is a framework building modular applications.
+//
+// It implements dependency injection using the fx libraryand adds on
```
```suggestion
// It implements dependency injection using the fx library and adds on
```

joamaki replied:
Fixed.

dylandreimerink reviewed (commented):
Some small nits, nothing major

joamaki commented on pkg/hive/cell/decorators.go:
```diff
@@ -0,0 +1,57 @@
+// Primary use-case for this helper is in operator which has two lifecycles:
+// an outer lifecycle for infrastructure and inner lifecycle for when it
+// is elected leader.
+func WithCustomLifecycle(lc fx.Lifecycle, cells ...Cell) Cell {
```
TODO: Remove, might as well do `cell.Decorate(func(lc MyCustomLifecycle) fx.Lifecycle { return lc }, ...)` instead.

christarazi reviewed (approved):
> There's now different cell types, so config isn't so baked in. The config cell now takes the default config as parameter, which can allow parametrizing the cell for different use-cases (e.g. the gops cell has different default port if it's used in agent vs operator).

Thanks, this helped ease my understanding a bit. 

Overall, it looks good to me. Maybe I'm overly pedantic about the `Fixes` in the commit msgs, but it would be really nice to keep chaining the `Fixes` so that we can properly track regressions such as the one fixed in the second commit and where they came from, especially with such a large refactor like moving to the hive / cell style code.

glibsm commented on pkg/hive/cell/cell.go:
```diff
@@ -0,0 +1,99 @@
+// Group creates a named group of cells. Groups are useful for passing
+// a set of constructors and invokes around. The name will be included
+// in the object dump (hive.PrintObjects) and in the dot graph (hive.PrintDotGraph).
+func Group(name string, cells ...Cell) Cell {
```
is there a reason to rebrand `fx.Module` to `Group` here?

joamaki replied:
Good point. I think I was somewhat stuck with not using it to avoid overlapping use, but now it makes sense since it's the same thing. I think I'm going to model this now much more closer to fx, with essentially just the extra `Config` thing on top.

glibsm commented on pkg/hive/cell/cell.go:
```diff
@@ -0,0 +1,99 @@
+//     specific set of cells.
+//   - Config: Cell providing a configuration struct.
+type Cell interface {
+	RegisterFlags(*pflag.FlagSet)
+	ToOption(settings map[string]any) (fx.Option, error)
```
This seems like two different interfaces, which we then can compose into one? Something like `io.Reader` and `io.Writer` making an `io.ReadWriter`

joamaki replied:
Not really two different interfaces as you can't do one without the other. You need to call `RegisterFlags` in order to be able to derive what settings to give to `ToOption`, so it doesn't make sense to separate them.

glibsm commented on pkg/hive/cell/cell.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+type In = fx.In
+type Out = fx.Out
```
missing godoc

glibsm commented on pkg/hive/cell/config.go:
```diff
@@ -0,0 +1,118 @@
+
+// CellConfig is implemented by configuration structs to provide configuration
+// for a cell.
+type CellConfig interface {
```
`cell.CellConfig` stutters. Just `cell.Config` should probably be sufficient here.

Same with all the other `Cell*` prefixed objects in this package

joamaki replied:
Ah yeah, looks like I was a bit blind to that. I'll rework the names a bit. `Config` is already used as the `ConfigCell` constructor (that struct doesn't need to be exported), so will need to come up with something else.

ti-mo replied:
`.New()` is more commonly used in Go for returning new objects, maybe that would fit? I might be missing the point.

joamaki replied:
`New` wouldn't fit as we've got 5 "cell constructors": `Config`, `Provide`, `Invoke`, `Decorate`, `Module`.

glibsm commented on pkg/hive/cell/config.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// ConfigCell is a cell for configuration.
+type ConfigCell struct {
```
aside from stuttering, this package contains a `ConfigCell` and `CellConfig` which could provide to be a confusing clash

glibsm commented on pkg/hive/cell/decorators.go:
```diff
@@ -0,0 +1,57 @@
+// a decorator cell.
+//
+// A decorator function is a function that takes as argument an object
+// in the hive and returns an object of the same type. The cells wrapped
```
Is there validation that the signature for the function is correct? This can be validate through reflection.

Perhaps in addition to reflection, we can change the signature from `any` to a decorator defined as `func[T any](o T) (T, error)`

joamaki replied:
Doesn't fx.Decorate validate it (eventually)? 

`func[T any](o T) (T, error)` seems much too narrow as often you want to take multiple objects. Ah, but I do need to fix the comment!

glibsm commented on pkg/hive/cell/helpers.go:
```diff
@@ -0,0 +1,90 @@
+//			return myStartFunc(a, b)
+//	  	})
+//	}))
+func OnStart(fn any) Cell {
```
I will probably recommend to keep the long form only, at least initially

My main concern is the output in the startup. For example, from the on start test:
```
    writer.go:40: [Fx] INVOKE		reflect.makeFuncStub() from module "test"
    writer.go:40: [Fx] INVOKE		reflect.makeFuncStub() from module "test"
```

joamaki replied:
I debated as well whether this was worth the cost of losing the function names, but felt it was worth it for use in tests and in command-line utilities like `cilium/cmd/preflight_k8s_valid_cnp.go`. The names can be recovered with a lifecycle hook that keeps track of the origin.

OTOH, even if this looks much prettier, maybe less is more at this point, plus better to expose the long form as it's easier to then understand what actually goes on. I'll remove this for now and then if it's possible to implement it correctly when this is directly on top of dig I can revisit.
